clients/src/main/java/org/apache/kafka/common/record/RecordBatch.java

```diff
@@ -0,0 +1,212 @@
+     * caution. Generally {@link #lastOffset()} is safer since access is efficient for all magic versions.
+     *
+     * @return The base offset of this message set (which may or may not be the offset of the first record
+     *         as described above).
+     */
+    long baseOffset();
+
+    /**
+     * Get the last offset in this record set (inclusive).
+     *
+     * @return The offset of the last record in this entry
+     */
+    long lastOffset();
+
+    /**
+     * Get the offset following this entry (i.e. the last offset contained in this entry plus one).
+     *
+     * @return the next consecutive offset following this entry
+     */
+    long nextOffset();
+
+    /**
+     * Get the message format version of this entry (i.e its magic value).
+     *
+     * @return the magic byte
+     */
+    byte magic();
+
+    /**
+     * Get the producer ID (PID) for this log entry. For older magic versions, this will return 0.
+     *
+     * @return The PID or 0 if there is none
+     */
+    long producerId();
+
+    /**
+     * Get the producer epoch for this log entry.
+     *
+     * @return The producer epoch, or 0 if there is none
+     */
+    short producerEpoch();
+
+    /**
+     * Get the first sequence number of this message set.
+     * @return The first sequence number
+     */
+    int baseSequence();
+
+    /**
+     * Get the last sequence number of this message set.
+     *
+     * @return The last sequence number
+     */
+    int lastSequence();
+
+    /**
+     * Get the compression type of this log entry
+     *
+     * @return The compression type
+     */
+    CompressionType compressionType();
+
+    /**
+     * Get the size in bytes of this entry, including the size of the record and the log overhead.
+     * @return The size in bytes of this entry
+     */
+    int sizeInBytes();
+
+    /**
+     * Get the count if it is efficiently supported by the message format (which is only the case
+     * for magic 2 and higher).
+     *
+     * @return The number of records in the batch or null for magic versions 0 and 1.
+     */
+    Integer countOrNull();
+
+    /**
+     * Check whether this entry contains a compressed message set.
+     * @return true if so, false otherwise
+     */
+    boolean isCompressed();
+
+    /**
+     * Write this entry into a buffer.
+     * @param buffer The buffer to write the entry to
+     */
+    void writeTo(ByteBuffer buffer);
+
+    /**
+     * Whether or not this log entry is part of a transaction.
+     * @return true if it is, false otherwise
+     */
+    boolean isTransactional();
+
+    /**
+     * Get the partition leader epoch of this entry.
+     * @return The leader epoch or -1 if it is unknown
+     */
+    int partitionLeaderEpoch();
+
+    /**
+     * A mutable log entry is one that can be modified in place (without copying).
+     */
+    interface MutableRecordBatch extends RecordBatch {
```

Comment:
Wouldn't this be better as a top level interface? I think an inner interface would make sense if we called it `Mutable` only. As it is, the name stands well on its own and it makes usage a bit less verbose.